site/en/blog/inside-the-container-query-polyfill/index.md

```diff
@@ -0,0 +1,301 @@
+
+{% Compare 'worse', 'before' %}
+```css
+@container (width > 300px) {
+  #foo::before {
+    /* ... */
+  }
+}
+```
+{% endCompare %}
+
+{% Compare 'better', 'after' %}
+```css
+@media all {
+  #foo:where([cq-XYZ~="123"])::before {
+    /* ... */
+  }
+}
+```
+{% endCompare %}
+
+Instead of being transformed to `#foo::before:where([cq-XYZ~="123"])` (which would be invalid), the conditional selector is moved to the end of the originating element, `#foo`.
+
+However, that’s not all that is needed. A container isn’t allowed to modify anything not contained _within_ it (and a container can’t be inside of itself), but consider that is exactly what would happen if `#foo` was itself the container element being queried. The `#foo[cq-XYZ]` attribute would be erroneously changed, and any `#foo` rules would be erroneously applied.
+
+To correct this, the polyfill actually uses _two_ attributes: one that can only be applied to an element by a parent, and one that an element can apply to itself. The latter attribute is used for selectors that target pseudo-elements.
+
+{% Compare 'worse', 'before' %}
+```css
+@container (width > 300px) {
+  #foo,
+  #foo::before {
+    /* ... */
+  }
+}
+```
+{% endCompare %}
+
+{% Compare 'better', 'after' %}
+```css
+@media all {
+  #foo:where([cq-XYZ-A~="123"]),
+  #foo:where([cq-XYZ-B~="123"])::before {
+    /* ... */
+  }
+}
+```
+{% endCompare %}
+
+Since a container will never apply the first attribute (`cq-XYZ-A`) to itself, the first selector will only match if a _different_ parent container has met the container conditions and applied it.
+
+### Container relative units
+
+Container queries also come with [a few new units](https://www.w3.org/TR/css-contain-3/#container-lengths) that you can use in your CSS, such as `cqw` and `cqh` for 1% of the width and height (respectively) of the closest appropriate parent container. To support these, the unit is transformed into a `calc(...)` expression using [CSS Custom Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*). The polyfill will set the values for these properties via inline styles on the container element.
+
+{% Compare 'worse', 'before' %}
+```css
+.card {
+  width: 10cqw;
+  height: 10cqh;
+}
+```
+{% endCompare %}
+
+{% Compare 'better', 'after' %}
+```css
+.card {
+  width: calc(10 * --cq-XYZ-cqw);
+  height: calc(10 * --cq-XYZ-cqh);
+}
+```
+{% endCompare %}
+
+There are also logical units, like `cqi` and `cqb` for inline size and block size (respectively). These are a little bit more complicated, because the inline and block axes are determined by the [`writing-mode`](https://developer.mozilla.org/en-US/docs/Web/CSS/writing-mode) of _the element using the unit_, not the element being queried. To support this, the polyfill applies an inline style to any element whose `writing-mode` differs from its parent.
```

Comment:
```suggestion
There are also logical units, like `cqi` and `cqb` for inline size and block size (respectively). These are a little bit more complicated, because the inline and block axes are determined by the [`writing-mode`](https://developer.mozilla.org/docs/Web/CSS/writing-mode) of _the element using the unit_, not the element being queried. To support this, the polyfill applies an inline style to any element whose `writing-mode` differs from its parent.
```